Downgrade AccessExclusiveLock's on system tables during PREPARE.

asimrp opened the pull request:
Keeping AccessExclusive locks on system tables across prepare and
commit phases of 2PC is not a good idea.  If we crash after prepare,
during recovery, we recreate the dummy PGPROC and all locks it held.
Letting the dummy PGPROC acquire locks on system tables in
AccessExclusive mode may block new backends from being initialized.
And that will not just be a deadlock but a point of no return.

It is safe to downgrade the lock during prepare because by then,
the transaction has finished writing data pages.  All that's left is
marking it committed.

asimrp commented:
Tests for this change cannot be added in installcheck as crash recovery and fail-over cycles need to exercised.  I'm covering the following scenarios in tests, using fault injectors in GPDB and TINC, the not-yet-open test framework.

1. Inject suspend fault so that QD sleeps after PREPARE phase. Acquire AccessExclusiveLock on catalog table in a transaction and commit, so that the fault is triggered.  From another transaction, try to read the same catalog table from QE.  The read should not be blocked.
2. Inject fault to crash QE after writing PREPARE record.  Acquire AccessExclusiveLock on a catalog table and commit the transaction to trigger the fault.  Recovery should go finish in reasonable time.  Without the fix, the QE will be perpetually stuck in recovery.
3. Inject a fault to fail over to mirror upon PREPARE.  The rest is same as test 2.

xiong-gang commented:
+1

danielgustafsson commented:
Can you update the commit message to include steps for how to test this for future reference? While steps related to TINC obviously don't make sense at least a high-level overview would be nice for when this ticket is long closed and forgotten and the commit message is our documentation.

atris commented:
+1

ashwinstar commented on src/backend/storage/lmgr/lock.c:
```diff
@@ -1999,14 +2001,56 @@ AtPrepare_Locks(void)
 			continue;
 
 		/*
+		 * Downgrade AccessExclusiveLock to ExclusiveLock on system
```
Just curious why not do it at original place where record.lockmode was set earlier after memcpy. Just seems better flow and aligns with the comment "Create a 2 PC record"

asimrp replied:
Thanks, that makes sense.  Let me update the pr.

asimrp commented:
On Wed, Mar 23, 2016 at 2:30 AM, Daniel Gustafsson <notifications@github.com
> wrote:

> Can you update the commit message to include steps for how to test this
> for future reference? While steps related to TINC obviously don't make
> sense at least a high-level overview would be nice for when this ticket is
> long closed and forgotten and the commit message is our documentation.
>
>
Yes, sure.  That makes sense.

ashwinstar commented:
:+1: Loved the thought process and approach Asim, on the fix to prevent this data unavailable and data loss scenario.

asimrp commented:
Updated commit message as suggested by Daniel and refactored as per Ashwin's suggestions.  I believe this fix is relevant to PostgreSQL as well.  E.g. tried this on 9.4_STABLE:

```
begin;
lock pg_database in access exclusive mode;
prepare transaction 'my_tx';
```
If we exit `psql` at this point, we are doomed.  We have a prepare transaction with AccessExclusiveLock help on `pg_database` and that is going to block any other backend from being initialized, even after restart.  Resetting xlog is the only way out.

hlinnaka commented:
On 23/03/16 20:23, Asim RP wrote:
> Updated commit message as suggested by Daniel and refactored as per Ashwin's suggestions.  I believe this fix is relevant to PostgreSQL as well.  E.g. tried this on 9.4_STABLE:
>
> ```
> begin;
> lock pg_database in access exclusive mode;
> prepare transaction 'my_tx';
> ```
> If we exit `psql` at this point, we are doomed.  We have a prepare
> transaction with AccessExclusiveLock help on `pg_database` and that
> is going to block any other backend from being initialized, even
> after restart.

Heh, that's a funny one :-).

I'm not immediately convinced that downgrading the lock is safe. At 
least in PostgreSQL, what if you had run VACUUM FULL on the table, which 
changes its relfilenode (because it works like CLUSTER or ALTER TABLE in 
modern versions of PostgreSQL)? Would it really be safe to let other 
processes to read the file, and which relfilenode would it read, and 
what would then happen when the prepared transaction commits or rolls 
back? In GPDB, I think REINDEX might have similar problems.

>  Resetting xlog is the only way out.

You can delete the file from pg_twophase to roll back the prepared 
transaction. It's still not nice, of course.

- Heikki

asimrp commented:
@hlinnaka, in GPDB, the old relfilenode of the index will still remain valid at PREPARE time.  By relinquishing AccessExclusiveLock, we are allowing other transactions to read this old relfile.  I could verify this by performing index lookup on e.g. `pg_attribute` while holding `REINDEX table pg_attribute` at the end of `AtPrepare_Locks()`.

In fact, we got to working on this PR because VACUUM FULL crashed right after prepare on a customer system.  Similar to REINDEX, the old relfilenode should remains valid at prepare time.  Can you elaborate the concern you have here?

ashwinstar commented:
deletes/updates are mvcc on pages but not drop / truncate of files as a
whole. So, operations doing the same would need to make sure no readers in
system to get rid of the files, hence believe need higher level of locking
for these scenarios. If reindex performed full operation as mvcc and didn't
physically delete the old file at end of transaction would have been safe
with just Exclusive. Sounds reasonable ?

On Tue, Mar 29, 2016 at 6:28 PM, Asim RP <notifications@github.com> wrote:

> @hlinnaka <https://github.com/hlinnaka>, in GPDB, the old relfilenode of
> the index will still remain valid at PREPARE time. By relinquishing
> AccessExclusiveLock, we are allowing other transactions to read this old
> relfile. I could verify this by performing index lookup on e.g.
> pg_attribute while holding REINDEX table pg_attribute at the end of
> AtPrepare_Locks().
>
> In fact, we got to working on this PR because VACUUM FULL crashed right
> after prepare on a customer system. Similar to REINDEX, the old relfilenode
> should remains valid at prepare time. Can you elaborate the concern you
> have here?
>
> —
> You are receiving this because you commented.
> Reply to this email directly or view it on GitHub
> <https://github.com/greenplum-db/gpdb/pull/562#issuecomment-203183805>
>

asimrp commented:
Agreed, it is not safe to downgrade AccessExclusiveLock after prepare, before commit.  Thank you Ashwin and Heikki.  Closing the PR.
